protocols/netlink/netlink_writer.go

```diff
@@ -0,0 +1,227 @@
+func (nw *NetlinkWriter) AddPath(pfx bnet.Prefix, path *route.Path) error {
+	// check if for this prefix already a route is existing
+	existingPaths, ok := nw.pt[pfx]
+
+	// if no route exists, add that route
+	if existingPaths == nil || !ok {
+		paths := make([]*route.Path, 1)
+		paths = append(paths, path)
+		nw.pt[pfx] = paths
+
+		// add the route to kernel
+		return nw.addKernel(pfx, path)
+	}
+
+	// if the new path is already in, don't do anything
+	for _, ePath := range existingPaths {
+		if ePath.Equal(path) {
+			return nil
+		}
+	}
+
+	existingPaths = append(existingPaths, path)
+	nw.pt[pfx] = existingPaths
+
+	// now add to netlink
+	return nw.addKernel(pfx, path)
+}
+
+// RemovePath removes a path from the Kernel using netlink This function is triggered by the loc_rib, cause we are subscribed as client in the loc_rib
+func (nw *NetlinkWriter) RemovePath(pfx bnet.Prefix, path *route.Path) bool {
+	// check if for this prefix already a route is existing
+	existingPaths, ok := nw.pt[pfx]
+
+	// if no route exists, nothing to do
+	if existingPaths == nil || !ok {
+		return true
+	}
+
+	// if the new path is already in: remove
+	removeIdx := 0
+	remove := false
+	for idx, ePath := range existingPaths {
+		if ePath.Equal(path) {
+			removeIdx = idx
+
+			remove = true
+			err := nw.removeKernel(pfx, path)
+			if err != nil {
+				log.WithError(err).Errorf("Error while removing path %s for prefix %s", path.String(), pfx.String())
+				remove = false
+			}
+
+			break
+		}
+	}
+
+	if remove {
+		existingPaths = append(existingPaths[:removeIdx], existingPaths[removeIdx+1:]...)
+		nw.pt[pfx] = existingPaths
+	}
+
+	return true
+}
+
+// Add pfx/path to kernel
+func (nw *NetlinkWriter) addKernel(pfx bnet.Prefix, path *route.Path) error {
+	route, err := nw.createRoute(pfx, path)
+	if err != nil {
+		log.Errorf("Error while creating route: %v", err)
+		return fmt.Errorf("Error while creating route: %v", err)
+	}
+
+	log.WithFields(log.Fields{
+		"Prefix": pfx.String(),
+		"Table":  route.Table,
+	}).Debug("AddPath to netlink")
+
+	err = netlink.RouteAdd(route)
+	if err != nil {
+		log.Errorf("Error while adding route: %v", err)
+		return fmt.Errorf("Error while adding route: %v", err)
+	}
+
+	return nil
+}
+
+// remove pfx/path from kernel
+func (nw *NetlinkWriter) removeKernel(pfx bnet.Prefix, path *route.Path) error {
+	log.WithFields(log.Fields{
+		"Prefix": pfx.String(),
+	}).Debug("Remove from netlink")
+
+	route, err := nw.createRoute(pfx, path)
+	if err != nil {
+		return fmt.Errorf("Error while creating route: %v", err)
+	}
+
+	err = netlink.RouteDel(route)
+	if err != nil {
+		return fmt.Errorf("Error while removing route: %v", err)
+	}
+
+	return nil
+}
+
+// create a route from a prefix and a path
+func (nw *NetlinkWriter) createRoute(pfx bnet.Prefix, path *route.Path) (*netlink.Route, error) {
+	if path.Type != route.NetlinkPathType {
+	}
+
+	switch path.Type {
```

Comment:
I'm not happy about the fact that we need to be protocol specific here.
Maybe we should add next-hop straight to Path?
@BarbarossaTM @czerwonk what do you guys think?

Reply:
We have to be at least a bit protocol specific.
Each route type has different parameters which should be considered to write into netlink. I don't think that we could do this in a generic way.
Therefore i thought it would be okay to have protocol specific code here which transforms a Path object of a certain type to a `netlink.Route` object

Reply:
as discussed: this is addressed in #133